classes/db/Db.php

```diff
@@ -202,6 +202,25 @@ abstract class DbCore
      */
     abstract public function getBestEngine();
 
+    /**
+     * Starts transaction
+     *
+     * @return bool true on success or false on failure
+     */
+    abstract public function beginTransaction(): bool;
```

Comment:
I think we should introduce a counter system to handle nested transactions, inspired by this example:
https://www.php.net/manual/fr/pdo.begintransaction.php#116669

Reply:
This class could be in charge of handling the counter, and call `protected doBeginTransaction` (instead of `parent::beginTransaction` in the example

This way the child classes only need to implement `protected doBeginTransaction` without worrying about the counter.

Reply:
The public methods from this class should also check the database engine, since transactions is only available with InnoDB engine, in case it's not the engine used the transaction can't and shouldn't be started

This will prevents shops using MyISAM databases to crash when transaction are used, of course they will be less performant and there will be risk of inconsistent states because rollback is not possible. But it's better than crashing I guess.

Reply:
One last thing! The transaction are auto-commited when we perform schema changing action like `DROP TABLE` `CREATE TABLE` `ALTER TABLE` or even `TRUNCATE TABLE`

So we should watch those actions and reset the counter when they happen, although it could result in errors later when `commit` or `rollback` are called, so I think we have two options in this case:

Reply:
- store the fact that an autocommit hase been forced by an action and fail silently (ignore the problem)

Reply:
An alternative would be to be able to configure which kind of error handling we prefer to have, and choose one as default

Reply:
is `MyISAM` still a thing these days? 🤔 

for the nested transactions etc. I'm not sure if we need such an implementation in legacy? I thought that this change supposed to help with one scenario where we could use just `Db::query('START TRANSACTION')` etc.

Reply:
Currently yes, we are still supporting MyISAM